### PR TITLE
Switch GHC 9.0.2 and GHC 9.2.1 to bignum-native

### LINKED ISSUE #153742: integer-simple does not exist for >= GHC 9.0
### Describe the bug

Nix provides `integer-simple` distributions for all versions. However, for GHC 9.0 integer-simple does not exist anymore. Instead it is replaced by `ghc-bignum`.

Currently the integer-simple distributions for GHC 9.0 in nixpkgs build and run but they are actually gmp distributions. 

### Steps To Reproduce

```
nix-shell -p "haskell.packages.integer-simple.ghc902.ghcWithPackages (pkgs: [])" --command "ghc-pkg list | grep gmp"
    integer-gmp-1.1
```
### Expected behavior

integer-simple distributions for GHC 9.0 and newer are replaced by distributions using the native GHC bignum backend. Note that this is configured using a different flag. A crude but working attempt which still uses the `integer-simple` infrastructure but changes the flag is the following but I expect we probably want a different attribute and flag:
```
-    INTEGER_LIBRARY = ${if enableIntegerSimple then "integer-simple" else "integer-gmp"}
+    BIGNUM_BACKEND = ${if enableIntegerSimple then "native" else "gmp"}
```

### Notify maintainers
@guibou @cdepillabout @expipiplus1 @maralorn @sternenseemann

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.0-91-generic, Ubuntu, 20.04.3 LTS (Focal Fossa)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4`
 - channels(moritz): `"nixpkgs-22.05pre334669.f366af7a1b3, nixpkgs-unstable-22.05pre334669.f366af7a1b3"`
 - nixpkgs: `/home/moritz/.nix-defexpr/channels/nixpkgs`

```

### PR DESCRIPTION
**Pull request by cocreature**
###### Motivation for this change

Neither of them support integer-simple anymore.

fixes #153742

###### Things done

Copied the infrastructure from ghcHead for bignum-native to 902 and 921 and dropped integer-simple (which did not work).

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by sternenseemann — APPROVED** [review]
Haven't checked build, but this is the diff checks out.

**Comment by cocreature** [comment]
I think I messed up something, it builds but I don’t end up with bignum-native. Investigating.

**Comment by cocreature** [comment]
Actually `ghcHEAD` also seems broken so that explains why copying still ends up with something broken
```
moritz@adjunction ~/c/nixpkgs (bignum-native)> NIX_PATH=nixpkgs=. nix-shell -p 'pkgs.haskell.packages.native-bignum.ghcHEAD.ghcWithPackages(pkgs: [])' --command "ghc-pkg list | grep gmp"
    integer-gmp-1.1
```
Did this use to work correctly?

**Comment by sternenseemann** [comment]
> Did this use to work correctly?

Binaries don't appear to be linking against gmp, so maybe the package db entry is to be expected?!

**Comment by cocreature** [comment]
Good point yeah apparently it is a compat library and it is expected. So it seems to work correctly for my ghc 902 package as well.

**Comment by cocreature** [comment]
Hey just wanted to make sure this isn’t blocked on me. If there are any more tests you’d like me to run before merging this let me know.

**Comment by sternenseemann** [comment]
Going to do a test build and merge later or tomorrow is the plan.

**Comment by sternenseemann** [comment]
Actually, can you adjust the commit message according to CONTRIBUTING.md? So prefixed with `haskell.compiler.ghc{902,921}: ` or similar.

**Comment by cocreature** [comment]
Adjusted the commit message

**Comment by cocreature** [comment]
Thanks @sternenseemann!

**Pull request merged**
